src/migrate/index.js

```diff
@@ -121,22 +131,83 @@ export default class Migrator {
     });
   }
 
+  _createMigrationLockTable(tableName) {
+    return this.knex.schema.createTable(tableName, function(t) {
+      t.integer('is_locked');
+    });
+  }
+
+  _getLockTableName(tableName) {
+    return tableName + '_lock';
+  }
+
+  _isLocked(trx) {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .transacting(trx)
+      .forUpdate()
+      .select('*')
+      .then(data => data[0].is_locked);
+  }
+
+  _lockMigrations(trx) {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .transacting(trx)
+      .update({ is_locked: 1 });
+  }
+
+  _unlockMigrations() {
+    var tableName = this._getLockTableName(this.config.tableName);
+    return this.knex(tableName)
+      .update({ is_locked: 0 });
+  }
+
   // Run a batch of current migrations, in sequence.
   _runBatch(migrations, direction) {
-    return Promise.all(_.map(migrations, this._validateMigrationStructure, this))
-      .then(() => this._latestBatchNumber())
-      .then((batchNo) => {
-        if (direction === 'up') batchNo++;
-        return batchNo;
-      })
-      .then((batchNo) => {
-        return this._waterfallBatch(batchNo, migrations, direction)
-      })
-      .catch((error) => {
-        helpers.warn('migrations failed with error: ' + error.message)
-        throw error
-      })
-  }
+    return this.knex.transaction(trx => {
+      return this._isLocked(trx)
+        .then(isLocked => {
+          if (isLocked) {
+            throw new LockError();
+          }
+        })
+        .then(() => this._lockMigrations(trx));
```

Comment:
Lets assume that here _lockMigrations throw an error, it wont be LockError so it will be passed until the end to normal error handler. But before reaching error handler there is still a chance that some other process did took lock correctly and now lock will be accidentally freed again :)
Some temporary connection error might trigger this. 
Probably the most correct functionality is that only the very same process who acquired the lock must be the one who releases it. So we should catch any exception thrown from transaction and throw LockError instead.